servers/audio/effects/audio_effect_hard_limiter.cpp

```diff
@@ -0,0 +1,168 @@
+		}
+
+		// Introduce latency by grabbing the AudioFrame stored previously,
+		// then overwrite it with current audioframe, then update circular
+		// buffer cursor.
+		float dst_buffer_left = sample_buffer_left[sample_cursor];
+		float dst_buffer_right = sample_buffer_right[sample_cursor];
+
+		sample_buffer_left[sample_cursor] = sample_left;
+		sample_buffer_right[sample_cursor] = sample_right;
+
+		sample_cursor = (sample_cursor + 1) % sample_buffer_left.size();
+
+		p_dst_frames[i].left = dst_buffer_left * gain;
+		p_dst_frames[i].right = dst_buffer_right * gain;
+	}
+}
+
+Ref<AudioEffectInstance> AudioEffectHardLimiter::instantiate() {
+	Ref<AudioEffectHardLimiterInstance> ins;
+	ins.instantiate();
+	ins->base = Ref<AudioEffectHardLimiter>(this);
+
+	float mix_rate = AudioServer::get_singleton()->get_mix_rate();
+
+	for (int i = 0; i < (int)Math::ceil(mix_rate * attack) + 1; i++) {
+		ins->sample_buffer_left.push_back(0.0f);
+		ins->sample_buffer_right.push_back(0.0f);
+	}
+
+	ins->gain_samples_to_store = (int)Math::ceil(mix_rate * (attack + sustain) + 1);
+	ins->gain_bucket_size = (int)(mix_rate * attack);
+
+	for (int i = 0; i < ins->gain_samples_to_store; i += ins->gain_bucket_size) {
+		ins->gain_buckets.push_back(1.0f);
+	}
+
+	return ins;
+}
+
+void AudioEffectHardLimiter::set_ceiling_db(float p_ceiling) {
+	ceiling = p_ceiling;
+}
+
+float AudioEffectHardLimiter::get_ceiling_db() const {
+	return ceiling;
+}
+
+float AudioEffectHardLimiter::get_pre_gain_db() const {
+	return pre_gain;
+}
+
+void AudioEffectHardLimiter::set_pre_gain_db(const float p_pre_gain) {
+	pre_gain = p_pre_gain;
+}
+
+float AudioEffectHardLimiter::get_release() const {
+	return release;
+}
+
+void AudioEffectHardLimiter::set_release(const float p_release) {
+	release = p_release;
+}
+
+void AudioEffectHardLimiter::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("set_ceiling_db", "ceiling"), &AudioEffectHardLimiter::set_ceiling_db);
+	ClassDB::bind_method(D_METHOD("get_ceiling_db"), &AudioEffectHardLimiter::get_ceiling_db);
+
+	ClassDB::bind_method(D_METHOD("set_pre_gain_db", "p_pre_gain"), &AudioEffectHardLimiter::set_pre_gain_db);
+	ClassDB::bind_method(D_METHOD("get_pre_gain_db"), &AudioEffectHardLimiter::get_pre_gain_db);
+
+	ClassDB::bind_method(D_METHOD("set_release", "p_release"), &AudioEffectHardLimiter::set_release);
+	ClassDB::bind_method(D_METHOD("get_release"), &AudioEffectHardLimiter::get_release);
+
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "pre_gain_db", PROPERTY_HINT_RANGE, "-24,24,0.01,suffix:dB"), "set_pre_gain_db", "get_pre_gain_db");
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "ceiling_db", PROPERTY_HINT_RANGE, "-24,0.0,0.01,suffix:dB"), "set_ceiling_db", "get_ceiling_db");
+	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "release", PROPERTY_HINT_RANGE, "0.01,3,0.01"), "set_release", "get_release");
+}
+
+AudioEffectHardLimiter::AudioEffectHardLimiter() {
+	pre_gain = 0.0;
+	ceiling = -0.3;
+	sustain = 0.02;
+	release = 0.1;
+}
```

Comment:
I would set those in the header file directly, so the constructor isn't needed.
Currently they get initialized twice (`0.0f` in header, and then custom values here).